TeaserImplTest & AdaptiveImageServletTest unit tests are failing with JDK12

Due to the changes in the reflection field access with JDK12 the unit tests TeaserImplTest and AdaptiveImageServletTest are failing.

Example output:
```
[ERROR] testWithInvalidDesignWidth  Time elapsed: 0.215 s  <<< ERROR!
java.lang.NoSuchFieldException: modifiers
	at com.adobe.cq.wcm.core.components.internal.servlets.AdaptiveImageServletTest.setFinalStatic(AdaptiveImageServletTest.java:637)
	at com.adobe.cq.wcm.core.components.internal.servlets.AdaptiveImageServletTest.testWithInvalidDesignWidth(AdaptiveImageServletTest.java:184)
```

--- a/bundles/core/pom.xml
+++ b/bundles/core/pom.xml
@@ -461,9 +461,12 @@
         <dependency>
             <groupId>org.javassist</groupId>
             <artifactId>javassist</artifactId>
-            <version>3.23.1-GA</version>
-            <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>uk.org.lidalia</groupId>
+            <artifactId>slf4j-test</artifactId>
+        </dependency>
+
 
         <!-- Only required for junit 4 -->
         <dependency>

--- a/bundles/core/src/test/java/com/adobe/cq/wcm/core/components/internal/models/v1/TeaserImplTest.java
+++ b/bundles/core/src/test/java/com/adobe/cq/wcm/core/components/internal/models/v1/TeaserImplTest.java
@@ -15,8 +15,6 @@
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
 package com.adobe.cq.wcm.core.components.internal.models.v1;
 
-import java.lang.reflect.Field;
-import java.lang.reflect.Modifier;
 import java.util.HashMap;
 import java.util.Objects;
 
@@ -25,11 +23,10 @@ import org.apache.sling.api.resource.ValueMap;
 import org.apache.sling.api.scripting.SlingBindings;
 import org.apache.sling.api.wrappers.ValueMapDecorator;
 import org.apache.sling.testing.mock.sling.servlet.MockSlingHttpServletRequest;
+import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.ExtendWith;
-import org.mockito.Mockito;
-import org.slf4j.Logger;
 
 import com.adobe.cq.sightly.WCMBindings;
 import com.adobe.cq.wcm.core.components.Utils;
@@ -40,9 +37,15 @@ import com.day.cq.commons.jcr.JcrConstants;
 import com.day.cq.wcm.api.components.Component;
 import io.wcm.testing.mock.aem.junit5.AemContext;
 import io.wcm.testing.mock.aem.junit5.AemContextExtension;
+import uk.org.lidalia.slf4jtest.TestLogger;
+import uk.org.lidalia.slf4jtest.TestLoggerFactory;
 
+import static org.hamcrest.Matchers.hasItem;
 import static org.junit.Assert.*;
-import static org.mockito.Mockito.*;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+import static uk.org.lidalia.slf4jtest.LoggingEvent.debug;
+import static uk.org.lidalia.slf4jtest.LoggingEvent.error;
 
 
 @ExtendWith(AemContextExtension.class)
@@ -67,25 +70,21 @@ class TeaserImplTest {
     private static final String TEASER_7 = TEST_ROOT_PAGE + TEST_ROOT_PAGE_GRID + "/teaser-7";
     private static final String TEASER_8 = TEST_ROOT_PAGE + TEST_ROOT_PAGE_GRID + "/teaser-8";
     private static final String TEASER_9 = TEST_ROOT_PAGE + TEST_ROOT_PAGE_GRID + "/teaser-9";
-    private Logger teaserLogger;
 
     private final AemContext context = CoreComponentTestContext.newAemContext();
+    private TestLogger testLogger;
 
     @BeforeEach
-    void setUp() throws IllegalAccessException, NoSuchFieldException {
+    void setUp() {
         context.load().json(TEST_BASE + CoreComponentTestContext.TEST_CONTENT_JSON, CONTENT_ROOT);
         context.load().json(TEST_BASE + CoreComponentTestContext.TEST_CONTENT_JSON, "/content/dam/core/images");
         context.load().binaryFile("/image/" + PNG_IMAGE_BINARY_NAME, PNG_ASSET_PATH + "/jcr:content/renditions/original");
+        testLogger = TestLoggerFactory.getTestLogger(TeaserImpl.class);
+    }
 
-        teaserLogger = Mockito.mock(Logger.class);
-        Field field = TeaserImpl.class.getDeclaredField("LOGGER");
-        Field modifiersField = Field.class.getDeclaredField("modifiers");
-        field.setAccessible(true);
-        // remove final modifier from field
-
-        modifiersField.setAccessible(true);
-        modifiersField.setInt(field, field.getModifiers() & ~Modifier.FINAL);
-        field.set(null, teaserLogger);
+    @AfterEach
+    void tearDown() {
+        TestLoggerFactory.clear();
     }
 
     @Test
@@ -128,29 +127,29 @@ class TeaserImplTest {
     @Test
     void testInvalidFileReference() {
         Teaser teaser = getTeaserUnderTest(TEASER_2);
-        verify(teaserLogger)
-                .error("Asset /content/dam/core/images/Adobe_Systems_logo_and_wordmark configured for the teaser component from /content/teasers/jcr:content/root/responsivegrid/teaser-2 doesn't exist.");
+        assertThat(testLogger.getLoggingEvents(), hasItem(error(
+                "Asset /content/dam/core/images/Adobe_Systems_logo_and_wordmark configured for the teaser component from /content/teasers/jcr:content/root/responsivegrid/teaser-2 doesn't exist.")));
         assertNull(teaser.getImageResource());
     }
 
     @Test
     void testEmptyFileReference() {
         Teaser teaser = getTeaserUnderTest(TEASER_3);
-        verify(teaserLogger)
-                .debug("Teaser component from /content/teasers/jcr:content/root/responsivegrid/teaser-3 does not have an asset or an image file configured.");
+        assertThat(testLogger.getLoggingEvents(), hasItem(debug(
+                "Teaser component from /content/teasers/jcr:content/root/responsivegrid/teaser-3 does not have an asset or an image file configured.")));
         assertNull(teaser.getImageResource());
     }
 
     @Test
     void testTeaserWithoutLink() {
         Teaser teaser = getTeaserUnderTest(TEASER_4);
-        verify(teaserLogger)
-                .debug("Teaser component from /content/teasers/jcr:content/root/responsivegrid/teaser-4 does not define a link.");
+        assertThat(testLogger.getLoggingEvents(),
+                hasItem(debug("Teaser component from /content/teasers/jcr:content/root/responsivegrid/teaser-4 does not define a link.")));
         assertNull(teaser.getLinkURL());
     }
 
     @Test
-    public void testTeaserWithHiddenLinks() {
+    void testTeaserWithHiddenLinks() {
         Teaser teaser = getTeaserUnderTest(TEASER_5,
                 Teaser.PN_TITLE_LINK_HIDDEN, true,
                 Teaser.PN_IMAGE_LINK_HIDDEN, true);

--- a/bundles/core/src/test/java/com/adobe/cq/wcm/core/components/internal/servlets/AdaptiveImageServletTest.java
+++ b/bundles/core/src/test/java/com/adobe/cq/wcm/core/components/internal/servlets/AdaptiveImageServletTest.java
@@ -21,8 +21,6 @@ import java.io.ByteArrayInputStream;
 import java.io.ByteArrayOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
-import java.lang.reflect.Field;
-import java.lang.reflect.Modifier;
 import javax.imageio.ImageIO;
 import javax.servlet.http.HttpServletResponse;
 
@@ -38,8 +36,6 @@ import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.ExtendWith;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
 
 import com.adobe.cq.wcm.core.components.internal.models.v1.AbstractImageTest;
 import com.adobe.cq.wcm.core.components.internal.models.v1.ImageImpl;
@@ -49,9 +45,13 @@ import com.day.cq.dam.api.handler.store.AssetStore;
 import com.day.cq.dam.commons.handler.StandardImageHandler;
 import com.day.image.Layer;
 import io.wcm.testing.mock.aem.junit5.AemContextExtension;
+import uk.org.lidalia.slf4jtest.TestLogger;
+import uk.org.lidalia.slf4jtest.TestLoggerFactory;
 
+import static org.hamcrest.Matchers.hasItem;
 import static org.junit.Assert.*;
 import static org.mockito.Mockito.*;
+import static uk.org.lidalia.slf4jtest.LoggingEvent.warn;
 
 @ExtendWith(AemContextExtension.class)
 class AdaptiveImageServletTest extends AbstractImageTest {
@@ -60,6 +60,7 @@ class AdaptiveImageServletTest extends AbstractImageTest {
 
     private AdaptiveImageServlet servlet;
     private static final int ADAPTIVE_IMAGE_SERVLET_DEFAULT_RESIZE_WIDTH = 1280;
+    private TestLogger testLogger;
 
     @BeforeEach
     void setUp() throws IOException {
@@ -73,12 +74,14 @@ class AdaptiveImageServletTest extends AbstractImageTest {
             return ImageIO.read(rendition.getStream());
         });
         servlet = new AdaptiveImageServlet(mockedMimeTypeService, assetStore, ADAPTIVE_IMAGE_SERVLET_DEFAULT_RESIZE_WIDTH);
+        testLogger = TestLoggerFactory.getTestLogger(AdaptiveImageServlet.class);
     }
 
     @AfterEach
     void tearDown() {
         resourceResolver = null;
         servlet = null;
+        TestLoggerFactory.clear();
     }
 
     @Test
@@ -180,8 +183,6 @@ class AdaptiveImageServletTest extends AbstractImageTest {
 
     @Test
     void testWithInvalidDesignWidth() throws Exception {
-        Logger logger = spy(LoggerFactory.getLogger("FakeLogger"));
-        setFinalStatic(AdaptiveImageServlet.class.getDeclaredField("LOGGER"), logger);
         Pair<MockSlingHttpServletRequest, MockSlingHttpServletResponse> requestResponsePair =
                 prepareRequestResponsePair(IMAGE1_PATH, "img.700", "png");
         MockSlingHttpServletRequest request = requestResponsePair.getLeft();
@@ -194,11 +195,8 @@ class AdaptiveImageServletTest extends AbstractImageTest {
         Dimension actualDimension = new Dimension(image.getWidth(), image.getHeight());
         assertEquals("Expected image rendered at requested size.", expectedDimension, actualDimension);
         assertEquals("Expected a PNG image.", "image/png", response.getContentType());
-        verify(logger).warn(
-                "One of the configured widths ({}) from the {} content policy is not a valid Integer.",
-                "invalid",
-                "/conf/$aem-mock$/settings/wcm/policies/core/wcm/components/image/v1/image/$mock-policy"
-        );
+        assertThat(testLogger.getLoggingEvents(), hasItem(warn("One of the configured widths ({}) from the {} content policy is not a " +
+                "valid Integer.", "invalid", "/conf/$aem-mock$/settings/wcm/policies/core/wcm/components/image/v1/image/$mock-policy")));
     }
 
     @Test
@@ -629,15 +627,6 @@ class AdaptiveImageServletTest extends AbstractImageTest {
         bindings.put(SlingBindings.RESOLVER, resourceResolver);
         request.setAttribute(SlingBindings.class.getName(), bindings);
         return new RequestResponsePair(request, response);
-    }
-
-    private void setFinalStatic(Field field, Object newValue) throws Exception {
-        field.setAccessible(true);
-        // remove final modifier from field
-        Field modifiersField = Field.class.getDeclaredField("modifiers");
-        modifiersField.setAccessible(true);
-        modifiersField.setInt(field, field.getModifiers() & ~Modifier.FINAL);
-        field.set(null, newValue);
     }
 
     private static class RequestResponsePair extends Pair<MockSlingHttpServletRequest, MockSlingHttpServletResponse> {

--- a/parent/pom.xml
+++ b/parent/pom.xml
@@ -757,6 +757,18 @@
                 <scope>test</scope>
             </dependency>
             <dependency>
+                <groupId>org.javassist</groupId>
+                <artifactId>javassist</artifactId>
+                <version>3.23.1-GA</version>
+                <scope>test</scope>
+            </dependency>
+            <dependency>
+                <groupId>uk.org.lidalia</groupId>
+                <artifactId>slf4j-test</artifactId>
+                <version>1.2.0</version>
+                <scope>test</scope>
+            </dependency>
+            <dependency>
                 <groupId>io.wcm</groupId>
                 <artifactId>io.wcm.testing.aem-mock.junit5</artifactId>
                 <version>2.7.0</version>

